### DISCUSSION TITLE
Formatting with black

### DISCUSSION THREAD

**Discussion by mostafa** [SigmaHQ/pySigma#129] [Ideas]
@thomaspatzke 

I suggest [black](https://github.com/psf/black) for formatting the entire codebase. Is it opinionated, yet produces visually clean code. Note that running black on the codebase kinda wipes the git history.

WDYT?

**Comment by thomaspatzke** [comment]
The git history argument kept me from doing this, but it [doesn't seems to be 100% valid anymore](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html), at least not for `git blame`, which is the most important, I believe.

On the other side I'm really annoyed by myself that I use different coding styles depending on mood and planet constellations 🤣

I think I will give it a try. What do you think about [gray](https://github.com/dizballanze/gray)? Could be a soft move, but I think it's more a compromise.

**Reply by mostafa** [reply]
I _just_ heard about gray and it seems interesting. It look much like [golangci-lint](https://github.com/golangci/golangci-lint). Yet, git-blame natively supports ignoring the history of certain commits, as already mentioned in black docs. From the looks of it, we have two options:

1. Using **Git-blame** with [blame.ignoreRevsFile](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt-blameignoreRevsFile) set, that can be used to ignore certain revision in the history and it receives a `.git-blame-ignore-revs` files that must contain commit SHAs. Apparently, [GitHub](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) added support for the feature back in 2022, which was in beta, but is now [available to everyone](https://github.com/orgs/community/discussions/5033#discussioncomment-2416240). It is documented [here](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view).
2. Using [**Git-black**](https://github.com/AlayaCare/git-black), which does the same thing in a different way.

I personally think that black leaves no room for errors, yet gray is backed by powerful QA tools. Either way, I can make a PR to see if the git-blame-history-feature actually works.

**Reply by mostafa** [reply]
@thomaspatzke 

I created a [PR](https://github.com/SigmaHQ/pySigma/pull/132) to fix this. I also added pre-commit hooks to run black before committing code locally.

**Comment by thomaspatzke** [comment]
[Invaluable knowledge](https://stackoverflow.com/questions/65304887/what-is-the-best-way-to-resolve-git-conflicts-after-reformatting-entire-python-c) if you try to rebase/merge larger changes after reformatting was committed.

**Comment by mostafa** [comment]
I suppose this can be closed now. Thanks for merging it.
